src/irmin.mli

```diff
@@ -1796,75 +1575,625 @@ module Private: sig
+        tree node or some file contents. It is [`Empty] if [k] is not
+        present in [t]. *)
 
-  include BC
+    val list: tree -> key -> (step * [`Contents | `Node]) list Lwt.t
+    (** [list t key] is the list of files and sub-nodes stored under [k]
+        in [t]. *)
 
-  module Key: Path.S with type t = key
-  (** [Key] provides base functions on paths. *)
+    (** {1 Diffs} *)
 
-  module Val: Contents.S with type t = value
-  (** [Val] provides base functions on user-defined, mergeable
-      contents. *)
+    val diff: tree -> tree -> (key * (contents * metadata) diff) list Lwt.t
+    (** [diff x y] is the difference of contents between [x] and [y]. *)
 
-  module Ref: Ref.S with type t = branch_id
-  (** [Ref] provides base functions on user-defined references. *)
+    (** {1 Manipulating Contents} *)
 
-  module Hash: Hash.S with type t = commit_id
-  (** [Hash] provides base functions on commit IDs. *)
+    val mem: tree -> key -> bool Lwt.t
+    (** [mem t k] is true iff [k] is associated to some contents in
+        [t]. *)
+
+    val findm: tree -> key -> (contents * metadata) option Lwt.t
+    (** [find t k] is [Some (b, m)] if [k] is associated to the contents
+        [b] and metadata [m] in [t] and [None] if [k] is not present in
+        [t]. *)
+
+    val find: tree -> key -> contents option Lwt.t
+    (** [find] is similar to {!find} but it discards metadata. *)
+
+    val getm: tree -> key -> (contents * metadata) Lwt.t
+    (** Same as {!find} but raise [Invalid_arg] if [k] is not present
+        in [t]. *)
+
+    val get: tree -> key -> contents Lwt.t
+    (** Same as {!get} but ignore the metadata. *)
+
+    val add: tree -> key -> ?metadata:metadata -> contents -> tree Lwt.t
+    (** [add t k c] is [t] where the key [k] is bound to the contents
+        [c]. *)
+
+    val remove: tree -> key -> tree Lwt.t
+    (** [remove t k] is [t] where [k] bindings has been removed. *)
+
+    (** {1 Manipulating Subtrees} *)
+
+    val memv: tree -> key -> bool Lwt.t
+    (** [memv t k] is false iff [getv k = `Empty]. *)
+
+    val getv: tree -> key -> tree Lwt.t
+    (** [getv t k] is [v] if [k] is associated to [v] in [t].  It is
+        [`Empty] if [k] is not present in [t]. *)
+
+    val addv: tree -> key -> tree -> tree Lwt.t
+    (** [addv t k v] is [t] where the key [k] is bound to the tree
+        [v]. *)
+
+    val merge: tree Merge.t
+    (** [merge] is the 3-way merge function for trees. *)
+
+    (** {1 Concrete Trees} *)
+
+    type concrete =
+      [ `Empty
+      | `Tree of (step * concrete) list
+      | `Contents of contents * metadata ]
+    (** The type for concrete trees. *)
+
+    val of_concrete: concrete -> tree
+    (** [of_concrete c] is the subtree equivalent to the concrete tree
+        [c]. *)
+
+    val to_concrete: tree -> concrete Lwt.t
+    (** [to_concrete t] is the concrete tree equivalent to the subtree
+        [t]. *)
+
+  end
+
+  (** {1 Reads} *)
```

Comment:
Perhaps all of these functions could be moved to a sub-module called `Not_threadsafe` or similar, to prevent people using them by accident?